lib/spack/spack/build_systems/cmake.py

```diff
@@ -355,6 +355,18 @@ def std_cmake_args(self):
         args += getattr(self.pkg, "cmake_flag_args", [])
         return args
 
+    @staticmethod
+    def setup_dependent_cmake_package(module, dependent_spec, *propegated_args):
+        """Helper method to propegate required CMake cache definitions to dependent
+        packages"""
+        if CMakeBuilder in type(dependent_spec.package.builder).__mro__:
+            dependent_std_args_stub = type(dependent_spec.package.builder).cmake_args
+
+            def new_cmake_args(self):
+                return propegated_args + dependent_std_args_stub(self)
+
+            type(dependent_spec.package.builder).cmake_args = new_cmake_args
```

Comment:
I'm somewhat worried about assigning the method this way: it could be difficult for someone in a debugging context to see that this was modified (other than stepping through functions)

Would it be possible to add a `dependency_cmake_args` property to `CMakePackage`? I think the most robust way to do that would be to explicitly acknowledge it in `cmake()`:

```
        options = self.std_cmake_args + self.dependency_cmake_args
        options += self.cmake_args()
```

As it's currently done, I'm also not sure how assigning the function combines with overrides in a subclass (I'm not sure that's a problem, but I'm having some trouble reasoning about it so I would want to check that carefully).

Reply:
> Would it be possible to add a dependency_cmake_args property to CMakePackage? I think the most robust way to do that would be to explicitly acknowledge it in cmake():

My first instinct is that in order to actually propagate this to the dependent classes, we need to make it visible to their modules, so we'd have to operate at the root `CMakeBuilder` scope, meaning, as I understand it, the implementation would have to look something like: 

```
CMakeBuilder:
...
    @staticmethod
    def dependency_args(*args):
        if not getattr(CMakeBuilder.dependency_args, "_extra_args"):
            _dependency_args._extra_args = []
        dependency_args._extra_args.extend(args)
...

    def cmake(...):
         ...
        options = self.std_cmake_args + CMakeBuilder.dependency_args._extra_args
```

Which, IMO is approximately as convoluted/susceptible to being overridden and broken.

W.r.t. the overridding, I think we'd need to make it clear that `setup_dependent_cmake_args` or `dependency_args` should not be overridden.


I'm going to try messing around with this implementation, maybe I've overthinking it

Reply:
Ok, I have a path forward I believe, implementing and pushing.